File: source/Imagen.cpp
#define STB_IMAGE_IMPLEMENTATION
#define STB_IMAGE_WRITE_IMPLEMENTATION

#include "Imagen.h"
#include "stb_image.h"
#include "stb_image_write.h"

#include <stdexcept>
#include <cmath>
#include <algorithm>

Imagen::Imagen() : _ancho(0), _alto(0) {}

Imagen::Imagen(const std::string& ruta) : _ancho(0), _alto(0) {
    cargar(ruta);
}

Imagen::~Imagen() {}

void Imagen::cargar(const std::string& ruta) {
    int canales;
    unsigned char* datos = stbi_load(ruta.c_str(), &_ancho, &_alto, &canales, 3);
    if (!datos) {
        throw std::runtime_error("No se pudo cargar la imagen: " + ruta);
    }

    _pixeles.assign(_alto, std::vector<std::vector<unsigned char>>(_ancho, std::vector<unsigned char>(3)));
    for (int f = 0; f < _alto; f++) {
        for (int c = 0; c < _ancho; c++) {
            int idx = (f * _ancho + c) * 3;
            _pixeles[f][c][0] = datos[idx];
            _pixeles[f][c][1] = datos[idx + 1];
            _pixeles[f][c][2] = datos[idx + 2];
        }
    }
    stbi_image_free(datos);
}

void Imagen::guardar(const std::string& ruta) const {
    if (estaVacia()) throw std::runtime_error("La imagen está vacía.");

    std::vector<unsigned char> datos(_alto * _ancho * 3);
    for (int f = 0; f < _alto; f++) {
        for (int c = 0; c < _ancho; c++) {
            int idx = (f * _ancho + c) * 3;
            datos[idx]     = _pixeles[f][c][0];
            datos[idx + 1] = _pixeles[f][c][1];
            datos[idx + 2] = _pixeles[f][c][2];
        }
    }

    // Detectar formato por extensión
    std::string ext = ruta.substr(ruta.find_last_of('.') + 1);
    std::transform(ext.begin(), ext.end(), ext.begin(), ::tolower);

    int ok = 0;
    if (ext == "png") {
        ok = stbi_write_png(ruta.c_str(), _ancho, _alto, 3, datos.data(), _ancho * 3);
    } else if (ext == "jpg" || ext == "jpeg") {
        ok = stbi_write_jpg(ruta.c_str(), _ancho, _alto, 3, datos.data(), 90);
    } else if (ext == "bmp") {
        ok = stbi_write_bmp(ruta.c_str(), _ancho, _alto, 3, datos.data());
    } else {
        throw std::runtime_error("Formato no soportado: " + ext + ". Usar .png, .jpg o .bmp");
    }

    if (!ok) throw std::runtime_error("No se pudo guardar la imagen en: " + ruta);
}

int Imagen::ancho() const { return _ancho; }
int Imagen::alto() const  { return _alto;  }

bool Imagen::estaVacia() const { return _ancho == 0 || _alto == 0; }

double Imagen::calcularEnergiaPixel(int fila, int col) const {
    auto pixel = [&](int f, int c) -> std::vector<unsigned char> {
        f = std::max(0, std::min(_alto - 1, f));
        c = std::max(0, std::min(_ancho - 1, c));
        return _pixeles[f][c];
    };

    double energia = 0.0;
    for (int canal = 0; canal < 3; canal++) {
        double dx = (double)pixel(fila, col + 1)[canal] - (double)pixel(fila, col - 1)[canal];
        double dy = (double)pixel(fila + 1, col)[canal] - (double)pixel(fila - 1, col)[canal];
        energia += dx * dx + dy * dy;
    }
    return std::sqrt(energia);
}

std::vector<std::vector<double>> Imagen::obtenerMatrizEnergia() const {
    std::vector<std::vector<double>> energia(_alto, std::vector<double>(_ancho));
    for (int f = 0; f < _alto; f++) {
        for (int c = 0; c < _ancho; c++) {
            energia[f][c] = calcularEnergiaPixel(f, c);
        }
    }
    return energia;
}

void Imagen::eliminarSeam(const std::vector<int>& seam) {
    if ((int)seam.size() != _alto) {
        throw std::runtime_error("El seam debe tener exactamente una entrada por fila.");
    }

    for (int f = 0; f < _alto; f++) {
        int col = seam[f];
        _pixeles[f].erase(_pixeles[f].begin() + col);
    }
    _ancho--;
}

File: source/python/Imagen.py
"""
Imagen.py — equivalente Python de Imagen.cpp
Usa Pillow para carga/guardado y NumPy para el cálculo de energía.

Las rutas de entrada/salida pueden ser str o pathlib.Path; ambas son aceptadas.
"""

import numpy as np
from pathlib import Path
from PIL import Image
import os


class Imagen:
    def __init__(self, ruta: str | Path | None = None):
        self._pixeles: np.ndarray | None = None  # shape (alto, ancho, 3)
        self._alto = 0
        self._ancho = 0
        if ruta is not None:
            self.cargar(ruta)

    # ── I/O ──────────────────────────────────────────────────────────────────

    def cargar(self, ruta: str | Path) -> None:
        ruta = Path(ruta).resolve()

        if not ruta.exists():
            raise FileNotFoundError(f"No se encontró la imagen: {ruta}")

        with Image.open(ruta) as img:
            img = img.convert("RGB")
            self._pixeles = np.array(img, dtype=np.uint8)

        self._alto, self._ancho = self._pixeles.shape[:2]

    def guardar(self, ruta: str | Path) -> None:
        if self.esta_vacia():
            raise RuntimeError("La imagen está vacía.")

        ruta = Path(ruta).resolve()
        ruta.parent.mkdir(parents=True, exist_ok=True)

        Image.fromarray(self._pixeles).save(ruta)

    # ── Propiedades ───────────────────────────────────────────────────────────

    def ancho(self) -> int:
        return self._ancho

    def alto(self) -> int:
        return self._alto

    def esta_vacia(self) -> bool:
        return self._ancho == 0 or self._alto == 0

    # ── Energía ───────────────────────────────────────────────────────────────

    def obtener_matriz_energia(self) -> list[list[float]]:
        if self._pixeles is None:
            raise RuntimeError("La imagen está vacía.")

        p = self._pixeles.astype(np.float64)

        # Gradiente horizontal
        p_right = p[:, np.minimum(np.arange(self._ancho) + 1, self._ancho - 1), :]
        p_left  = p[:, np.maximum(np.arange(self._ancho) - 1, 0), :]
        dx = p_right - p_left

        # Gradiente vertical
        rows = np.arange(self._alto)
        p_down = p[np.minimum(rows + 1, self._alto - 1), :, :]
        p_up   = p[np.maximum(rows - 1, 0), :, :]
        dy = p_down - p_up

        energia_arr = np.sqrt((dx**2 + dy**2).sum(axis=2))

        return energia_arr.tolist()

    # ── Eliminación de seam ───────────────────────────────────────────────────

    def eliminar_seam(self, seam: list[int]) -> None:
        if self._pixeles is None or len(seam) != self._alto:
            raise RuntimeError("El seam debe tener exactamente una entrada por fila.")

        if self._ancho <= 1:
            raise RuntimeError("No se puede eliminar seam: ancho <= 1.")

        nuevo = np.zeros((self._alto, self._ancho - 1, 3), dtype=np.uint8)

        for f in range(self._alto):
            col = seam[f]

            if col < 0 or col >= self._ancho:
                raise ValueError(f"Índice inválido en seam: fila {f}, columna {col}")

            nuevo[f, :col, :] = self._pixeles[f, :col, :]
            nuevo[f, col:, :] = self._pixeles[f, col + 1:, :]

        self._pixeles = nuevo
        self._ancho -= 1